/* Copyright (C) 2022. Huawei Technologies Co., Ltd. All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the Apache License Version 2.0.You may not use this file except in compliance with the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Apache License for more details at
 * http://www.apache.org/licenses/LICENSE-2.0
 */
syntax = "proto3";

package dataflow;
import "ge_ir.proto";

message FlowEdge {
  string node_name = 1; // node name
  uint32 index = 2; // node input/output index
}

message ProcessFunc {
  string name = 1; // Func name
  repeated uint32 in_index = 2; // Func inputs
  repeated uint32 out_index = 3; // Func outputs
}

message ProcessPoint {
  enum ProcessPointType {
    FUNCTION = 0;
    GRAPH = 1;
    INNER = 2;
    INVALID = 3;
  }

  string name = 1; // PP name
  ProcessPointType type = 2; // PP type : graph, func
  string compile_cfg_file = 3;
  repeated ProcessFunc funcs = 4; // when type is func pp
  repeated string graphs = 5; // when type is graph pp
  map<string, ProcessPoint> invoke_pps = 6; // pp invoke pps
  repeated FlowEdge in_edges = 7; // node inputs
  repeated FlowEdge out_edges = 8; // node outputs
  map<string, ge.proto.AttrDef> attrs = 9; // func init attrs
  bool is_built_in = 10;
  map<string, string> pp_extend_attrs = 11; // pp extend attrs
}
